src/libData/BlockChainData/BlockChain.h

```diff
@@ -0,0 +1,153 @@
+* warranties are given as to title or non-infringement, merchantability or fitness for purpose 
+* and, to the extent permitted by law, all liability for your use of the code is disclaimed. 
+* Some programs in this code are governed by the GNU General Public License v3.0 (available at 
+* https://www.gnu.org/licenses/gpl-3.0.en.html) (‘GPLv3’). The programs that are governed by 
+* GPLv3.0 are those programs that are located in the folders src/depends and tests/depends 
+* and which include a reference to GPLv3 in their program files.
+**/
+
+#ifndef __BLOCKCHAIN_H__
+#define __BLOCKCHAIN_H__
+
+#include <mutex>
+#include <vector>
+
+#include <boost/multiprecision/cpp_int.hpp>
+
+#include "libData/BlockData/Block/DSBlock.h"
+#include "libData/DataStructures/CircularArray.h"
+#include "libPersistence/BlockStorage.h"
+
+/// Transient storage for DS blocks.
+template<class T> class BlockChain
+{
+    std::mutex m_mutexBlocks;
+    CircularArray<T> m_blocks;
+
+public:
+    /// Constructor.
+    BlockChain()
+    {
+        m_blocks.resize(is_same<T, DSBlock>::value ? DS_BLOCKCHAIN_SIZE
+                                                   : TX_BLOCKCHAIN_SIZE);
+    }
+
+    /// Destructor.
+    ~BlockChain() {}
+
+    /// Reset
+    void Reset()
+    {
+        m_blocks.resize(is_same<T, DSBlock>::value ? DS_BLOCKCHAIN_SIZE
+                                                   : TX_BLOCKCHAIN_SIZE);
+    }
+
+    /// Returns the number of blocks.
+    boost::multiprecision::uint256_t GetBlockCount()
+    {
+        lock_guard<mutex> g(m_mutexBlocks);
+        return m_blocks.size();
+    }
+
+    /// Returns the last stored block.
+    T GetLastBlock()
+    {
+        lock_guard<mutex> g(m_mutexBlocks);
+        return m_blocks.back();
+    }
+
+    /// Returns the block at the specified block number.
+    TxBlock GetTxBlock(const boost::multiprecision::uint256_t& blockNum)
+    {
+        lock_guard<mutex> g(m_mutexBlocks);
+
+        if (blockNum >= m_blocks.size())
+        {
+            throw "Blocknumber Absent";
+        }
+        else if (blockNum + m_blocks.capacity() < m_blocks.size())
+        {
+            TxBlockSharedPtr block;
+            BlockStorage::GetBlockStorage().GetTxBlock(blockNum, block);
+            return *block;
+        }
+
+        // To-do: We cannot even index into a vector using uint256_t
+        // uint256_t might just be too big to begin with
+        // Consider switching to uint64_t
+        // For now we directly cast to uint64_t
+
+        if (m_blocks[blockNum].GetHeader().GetBlockNum() != blockNum)
+        {
+            LOG_GENERAL(FATAL,
+                        "assertion failed (" << __FILE__ << ":" << __LINE__
+                                             << ": " << __FUNCTION__ << ")");
+        }
+
+        return m_blocks[blockNum];
+    }
+
+    DSBlock GetDSBlock(const boost::multiprecision::uint256_t& blockNum)
```

Comment:
It's impossible to use this API correctly in a threaded environment (unless we have further context), for the reasons explained above.